### PR TITLE
(feat)(core): Add traffic congestion segments to the main Route

### PR DESCRIPTION
**Pull request by engali94**
This approach involves:
1. Adding a property or method to the main Route struct to retrieve congestion segments.
2. Each adapter would be responsible for extracting and mapping its specific congestion data to a common format when converting from the adapter-specific route to the main Route.

Pros:
- Keeps the congestion data close to the route data it belongs to.
- Provides a unified interface for accessing congestion data across different platforms.

Cons:
- Slightly increases the complexity of the core Route struct.

This is the implementation of  #292

### CODE REVIEW

**Review by ianthetechie — COMMENTED** [review]
Thanks for the PR @engali94!

I think @Archdoog is currently working on the best way(s) to expose annotations up to users (maybe that should be an RFD? 😉). I _think_ that our current ideal scenario is to not make the route struct larger unless a field is likely to be supported by the vast majority of likely implementations. For example, we made the decision to expose voice and banner instructions at the step level  because this is a de facto standard now (though it's still optional).

I'm not sure if traffic congestion warrants that level of treatment, but we DO want to get common data models and processing code into the codebase. I think the current idea is to push the whole set of annotations up _at the step level_ (not the full route level, since usually this is more actionable at the step level, but apps could reshape this into a route-long traffic layer after fetching a route, for example, to preserve this across the trip even when steps are dropped).

Annotations at the step level have landed as of #261. #287 was just opened yesterday, and shows the approach that @Archdoog currently has in mind, which is to transform `AnyAnnotation`s in Swift/Kotlin (still TODO). We'll provide the basic types in library code, as well as common annotation formats (ex: Valhalla). The beauty is that if your own server does something different, you just write your own struct (using the shared models, like speed limit and traffic congestion + anything that may be proprietary).

TL;DR, I think we can, with a bit of discussion, get all of the pros without the cons (which you correctly identified in your description :D).

Hopefully that makes sense, thanks for reading this tome, and thanks in advance for your patience as this part of the code is still in flux on our side :)

File: common/ferrostar/src/routing_adapters/osrm/utilities.rs [hunk 1]
```diff
@@ -47,12 +49,93 @@ pub(crate) fn zip_annotations(annotation: AnyAnnotation) -> Vec<AnyAnnotationVal
...
+    let lat2 = coord2.lat.to_radians();
+    let delta_lat = (coord2.lat - coord1.lat).to_radians();
+    let delta_lon = (coord2.lng - coord1.lng).to_radians();
+
+    let a =
+        (delta_lat / 2.0).sin().powi(2) + lat1.cos() * lat2.cos() * (delta_lon / 2.0).sin().powi(2);
+    let c = 2.0 * a.sqrt().atan2((1.0 - a).sqrt());
+
+    EARTH_RADIUS * c
+}
```
[the remark this answers is not in the document]
Comment (engali94) [reply common/ferrostar/src/routing_adapters/osrm/utilities.rs:131]: hi @ianthetechie I appreciate you very detailed review, I am very willing to discuss it further 
do you have any better/faster mean of communication to discuss this further?

[the remark this answers is not in the document]
Comment (ianthetechie) [reply common/ferrostar/src/routing_adapters/osrm/utilities.rs:131]: @engali94 yes! Feel free to reach out on Slack (link in the README); feel free to DM (Ian Wagner).

[the remark this answers is not in the document]
Comment (ianthetechie) [reply common/ferrostar/src/routing_adapters/osrm/utilities.rs:131]: @engali94 quick bump here to let you know that I'm available on Slack either via DM or in the `#ferrostar` channel. Not sure if this works for you time zone-wise, but we also have a regular meeting at 1:00 UTC on Tuesdays (Monday 6pm Pacific / 10am Seoul) to review PRs and discuss direction. Shoot me a message if you'd like to join.

**Comment by Archdoog** [comment]
I've created the following tickets to assist with wrapping annotations (including congestion up):

#317 
#316 
#320 

@engali94 feel free to comment on those issues and close this PR if they satisfy what you need plus #287

**Comment by engali94** [comment]
closing in favor of #287
